src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -190,16 +182,37 @@ internal static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationPro
                 protocolSize += protocol.Protocol.Length + 1;
             }
 
-            byte[] buffer = new byte[protocolSize];
-            var offset = 0;
+            return protocolSize;
+        }
+
+        internal static void SerializeAlpnProtocolList(List<SslApplicationProtocol> applicationProtocols, Span<byte> buffer)
+        {
+            Debug.Assert(GetAlpnProtocolListSerializedLength(applicationProtocols) == buffer.Length,
+                "GetAlpnProtocolListSerializedSize(applicationProtocols) == buffer.Length");
+
+            int offset = 0;
             foreach (SslApplicationProtocol protocol in applicationProtocols)
             {
-                buffer[offset++] = (byte)(protocol.Protocol.Length);
-                protocol.Protocol.Span.CopyTo(buffer.AsSpan(offset));
+                buffer[offset++] = (byte)protocol.Protocol.Length;
+                protocol.Protocol.Span.CopyTo(buffer.Slice(offset));
                 offset += protocol.Protocol.Length;
             }
+        }
+
+        internal static unsafe int SslSetAlpnProtos(SafeSslHandle ssl, List<SslApplicationProtocol> applicationProtocols)
+        {
+            int length = GetAlpnProtocolListSerializedLength(applicationProtocols);
+            Span<byte> buffer = length <= 256 ? stackalloc byte[256].Slice(0, length) : new byte[length];
+            SerializeAlpnProtocolList(applicationProtocols, buffer);
+            return SslSetAlpnProtos(ssl, buffer);
+        }
 
-            return buffer;
+        internal static unsafe int SslSetAlpnProtos(SafeSslHandle ssl, Span<byte> serializedProtocols)
+        {
+            fixed (byte* pBuffer = &MemoryMarshal.GetReference(serializedProtocols))
```

Comment:
```suggestion
            fixed (byte* pBuffer = serializedProtocols)
```

Reply:
I was discussing this with @MihaZupan when fixing his comment (https://github.com/dotnet/runtime/pull/63674#discussion_r782930142), the MemoryMarshal version apparently generates shorter code as it does not insert a Length check [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6mpoDMTUgwAqqjEMl5cNAN40G/pntmJAZuXGwAMxgmFAYAQVwABQheYygAClkxbC4AHmAATwwYAD4GQuLcAEo/AN9qAMaGAHpmhgA5CAwGcKjQrmlcXHKi1QZsWHHJWGwAEwKGCS4uGFnapoZZDA4IiPT0ipgAKiqAVS5ewQTk1OKM2AiGAFkYfGgCp4ncAAtsSRYAOIwDBsGBRWBcFQHUbVKpVDQNAIAX3WDFRxCC5BCYUi0WIsQAYrwEKtMtk8ocyodqqj6ht/BFiasGNDikdFik0gwALwjSo1REbOn0xpbHZ7MScu7w1GNJEi1EowVo5UY7TYi64mIMIkk2akMk5fKjKkwgWNYUbRl6lmHdmS24yHkMABkLzeUA+X1+/yBILBMhgkJgrNUcNldQjGzFu3SDrSMuVcoVyqVjVRAG0XhhvhBZgZxJJ0tnc/nCwB5MR8CAXFidIySVKpADmVQAuuisf1LtqY3s7RzHVBzZGkwFFTQkUA===)